invokeai/frontend/web/src/features/nodes/util/graphBuilders/addHrfToGraph.ts

```diff
@@ -0,0 +1,210 @@
+import { RootState } from 'app/store/store';
+import { NonNullableGraph } from 'features/nodes/types/types';
+import {
+  DenoiseLatentsInvocation,
+  RescaleLatentsInvocation,
+  MetadataAccumulatorInvocation,
+  NoiseInvocation,
+  LatentsToImageInvocation,
+  Edge,
+} from 'services/api/types';
+import {
+  LATENTS_TO_IMAGE,
+  DENOISE_LATENTS,
+  NOISE,
+  MAIN_MODEL_LOADER,
+  METADATA_ACCUMULATOR,
+  LATENTS_TO_IMAGE_HRF,
+  DENOISE_LATENTS_HRF,
+  RESCALE_LATENTS,
+  NOISE_HRF,
+  VAE_LOADER,
+} from './constants';
+
+// Copy certain connections from previous DENOISE_LATENTS to new DENOISE_LATENTS_HRF.
+function copyConnectionsToDenoiseLatentsHrf(graph: NonNullableGraph): void {
+  const destinationFields = [
+    'vae',
+    'control',
+    'ip_adapter',
+    'metadata',
+    'unet',
+    'positive_conditioning',
+    'negative_conditioning',
+  ];
+  const newEdges: Edge[] = [];
+
+  // Loop through the existing edges connected to DENOISE_LATENTS
+  graph.edges.forEach((edge: Edge) => {
+    if (
+      edge.destination.node_id === DENOISE_LATENTS &&
+      destinationFields.includes(edge.destination.field)
+    ) {
+      // Add a similar connection to DENOISE_LATENTS_HRF
+      newEdges.push({
+        source: {
+          node_id: edge.source.node_id,
+          field: edge.source.field,
+        },
+        destination: {
+          node_id: DENOISE_LATENTS_HRF,
+          field: edge.destination.field,
+        },
+      });
+    }
+  });
+  graph.edges = graph.edges.concat(newEdges);
+}
+
+// Adds the high-res fix feature to the given graph.
+export const addHrfToGraph = (
+  state: RootState,
+  graph: NonNullableGraph
+): void => {
+  const { vae } = state.generation;
+  const isAutoVae = !vae;
+
+  // Pre-existing (original) graph nodes.
+  const originalDenoiseLatentsNode = graph.nodes[
+    DENOISE_LATENTS
+  ] as DenoiseLatentsInvocation;
+  const originalNoiseNode = graph.nodes[NOISE] as NoiseInvocation;
+  const originalLatentsToImageNode = graph.nodes[
+    LATENTS_TO_IMAGE
+  ] as LatentsToImageInvocation;
+
+  // Scale height and width by hrfScale.
+  const hrfScale = state.generation.hrfScale;
+  const scaledHeight = originalNoiseNode?.height
+    ? originalNoiseNode.height * hrfScale
+    : undefined;
+  const scaledWidth = originalNoiseNode?.width
+    ? originalNoiseNode.width * hrfScale
+    : undefined;
```

Comment:
Back in v2.3, we handled output dimensions from the opposite direction. The width and height would represent the final output size, and then we would calculate the best starting dimensions from that. 

It would involve mutating that original noise node to have the calculated size, and would let you do away with the `hrfScale` parameter.

I think that provides a more intuitive UX than selecting the scale.

Reply:
We could allow the user to pass in the initial height and width of the HRF generation, with the max height/width of the slider being the current height/width, with reflected aspect ratio locking behavior. This would allow the user to play with the trade off between greater detail with generating closer to the native resolution and less HR artifacts by generating closer to the trained resolution.

What do you think?